lib/pengine/tests/native/native_find_rsc_test.c

```diff
@@ -0,0 +1,683 @@
+    assert_ptr_equal(mysql_group_1, native_find_rsc(mysql_clone_group, "mysql-group:1", cluster01, pe_find_current));
+    assert_ptr_equal(mysql_group_1, native_find_rsc(mysql_clone_group, "mysql-group", cluster01, pe_find_any|pe_find_current));
+    assert_ptr_equal(mysql_group_1, native_find_rsc(mysql_clone_group, "mysql-group", cluster01, pe_find_anon|pe_find_current));
+}
+
+static void
+clone_group_member_rsc(void **state) {
+    pe_resource_t *mysql_proxy = NULL;
+
+    /* Find the "mysql-proxy" resource, a member of "mysql-group". */
+    for (GList *iter = mysql_clone_group->children; mysql_proxy == NULL && iter != NULL; iter = iter->next) {
+        pe_resource_t *rsc = (pe_resource_t *) iter->data;
+
+        if (strcmp(rsc->id, "mysql-group:0") == 0) {
+            for (GList *iter2 = rsc->children; iter2 != NULL; iter2 = iter2->next) {
+                pe_resource_t *child = (pe_resource_t *) iter2->data;
+
+                if (strcmp(child->id, "mysql-proxy:0") == 0) {
+                    mysql_proxy = child;
+                    break;
+                }
+            }
+        }
+    }
+
+    assert_non_null(mysql_proxy);
+
+    /* Passes because NULL was passed for node, regardless of flags. */
+    assert_ptr_equal(mysql_proxy, native_find_rsc(mysql_proxy, "mysql-proxy:0", NULL, 0));
+    assert_ptr_equal(mysql_proxy, native_find_rsc(mysql_proxy, "mysql-proxy:0", NULL, pe_find_current));
+
+    /* Passes because resource's parent is a clone. */
+    assert_ptr_equal(mysql_proxy, native_find_rsc(mysql_proxy, "mysql-proxy:0", NULL, pe_find_clone));
+    assert_ptr_equal(mysql_proxy, native_find_rsc(mysql_proxy, "mysql-proxy:0", cluster02, pe_find_clone|pe_find_current));
+
+    /* Fails because mysql-proxy:0 is not running on cluster01, even with the right flags. */
+    assert_null(native_find_rsc(mysql_proxy, "mysql-proxy:0", cluster01, pe_find_current));
+
+    /* Fails because pe_find_current is required if a node is given. */
+    assert_null(native_find_rsc(mysql_proxy, "mysql-proxy:0", cluster02, 0));
+
+    /* Passes because mysql-proxy:0 is running on cluster02. */
+    assert_ptr_equal(mysql_proxy, native_find_rsc(mysql_proxy, "mysql-proxy:0", cluster02, pe_find_current));
+}
+
+int main(int argc, char **argv) {
+    const struct CMUnitTest tests[] = {
+        cmocka_unit_test(bad_args),
+        cmocka_unit_test(primitive_rsc),
+        cmocka_unit_test(group_rsc),
+        cmocka_unit_test(inactive_group_rsc),
+        cmocka_unit_test(group_member_rsc),
+        cmocka_unit_test(inactive_group_member_rsc),
+        cmocka_unit_test(clone_rsc),
+        cmocka_unit_test(inactive_clone_rsc),
+        cmocka_unit_test(clone_instance_rsc),
+        cmocka_unit_test(renamed_rsc),
+        cmocka_unit_test(bundle_rsc),
+        cmocka_unit_test(bundle_replica_rsc),
+        cmocka_unit_test(clone_group_rsc),
+        cmocka_unit_test(clone_group_instance_rsc),
+        cmocka_unit_test(clone_group_member_rsc),
```

Comment:
Add a todo comment that we should also test finding on allocated node (passing a node without pe_find_current, after scheduling has been done, for a resource that's starting, stopping, or moving). It's not worth the trouble right now ...

Reply:
Added.